### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by unorjikingsley**
## The following instructions was implemented in this PR:

- [X] Each class from the [UML class diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) should be created in a separate .rb file.
- [X] All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
- [X] All properties visible in the diagram should be defined and set up in the constructor method. Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created.
- [X] All methods visible in the diagram should be implemented:
     > - add_item method (in all classes that have that method)
         1. should take an instance of the Item class as an input.
         2. should add the input item to the collection of items.
         3. should add self as a property of the item object (by using the correct setter from the item object).
    > - can_be_archived?() in the Item class
         1. should return true if published_date is older than 10 years.
         2. otherwise, it should return false.
    > - move_to_archive() in the Item class
         1. should reuse can_be_archived?() method.
         2. should change the archived property to true if the result of the can_be_archived?() method is true.
         3. should do nothing if the result of the can_be_archived?() method is false.
    > - can_be_archived?() in the Book class
        1. should override the method from the parent class.
        2. should return true if parent's method returns true OR if cover_state equals to "bad".
        3. otherwise, it should return false.
    > - can_be_archived?() in the MusicAlbum class
       1. should override the method from the parent class.
       2. should return true if parent's method returns true AND if on_spotify equals true.
       3. otherwise, it should return false.
    > - can_be_archived?() in the Movie class
       1. should override the method from the parent class.
       2. should return true if parent's method returns true OR if silent equals true.
       3. otherwise, it should return false.
    > - can_be_archived?() in the Game class
       1. should override the method from the parent class.
       2. should return true if parent's method returns true AND if last_played_at is older than 2 years.
       3. otherwise, it should return false.
    > - Add unit tests for all implemented methods.

**User interface**

- [X] Create a main.rb file that will serve as your console app entry-point.
- [X] Your console app, at the start, should:
    > - Present the user with a list of options to perform.
    > - Let users choose an option.
    > - If needed, ask for parameters for the option.
    > - Have a way to quit the app.
- [X] The following options should be available:
    > - List all books
    > - List all music albums
    > - List all movies
    > - List of games
    > - List all genres (e.g 'Comedy', 'Thriller')
    > - List all labels (e.g. 'Gift', 'New')
    > - List all authors (e.g. 'Stephen King')
    > - List all sources (e.g. 'From a friend', 'Online shop')
    > - Add a book
    > - Add a music album
    > - Add a movie
    > - Add a game
- [X] All data should be preserved by saving collections in .json files.

**Database**
- [X] Create a schema.sql file with tables that will be analogical to the structure of the classes in your app. As you cannot implement inheritance in the database tables - add all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.

### CODE REVIEW

File: schema/schema.sql [hunk 1]
```diff
@@ -0,0 +1,75 @@
+CREATE DATABASE catalogue_of_things;
+
+CREATE TABLE books (
```
Comment (muneebulrehman) [inline schema/schema.sql:3]: The book should have a reference to the author as well as the genre. Kindly create the necessary foreign keys and columns with corresponding indexes.

File: schema/schema.sql [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+CREATE TABLE genre
+(
+    id INT PRIMARY KEY NOT NULL,
+    name VARCHAR(100),
+    items TEXT
+    []
+);
+    
+    CREATE TABLE music_album
```
Comment (muneebulrehman) [inline schema/schema.sql:46]: The music album should have a reference to the author as well as the label. Kindly create the necessary foreign keys and columns with corresponding indexes.

File: schema/schema.sql [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    (
+        id INT PRIMARY KEY NOT NULL,
+        on_spotify BOOLEAN,
+        publish_date DATE,
+        archived BOOLEAN,
+        genre_id INT REFERENCES genre(id)
+    );
+
+    CREATE TABLE games (
```
Comment (muneebulrehman) [inline schema/schema.sql:55]: The book should have a reference to the label as well as the genre. Kindly create the necessary foreign keys and columns with corresponding indexes.

**Review by muneebulrehman — CHANGES_REQUESTED** [review]
Hi team, 

Good job so far! I see you put a lot of effort into this project. However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- No linter errors
- Descriptive PR
- Great job with saving the data
- Contribution by all partners.

## Required Changes ♻️
_Check the comments under the review._

- [ ] The max length of the video is 5 min. Kindly re-record the video with a max length of 5 minutes. The whole team should talk about their code and contribution instead of one person walking through the whole project. And each member should highlight one piece of code they are proud of.
- [ ] The tests for the `item` class are missing. Kindly add the tests for the same before submitting it for re-review. Also, make sure, all the public methods in all classes should have unit tests.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: classes/book/book.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  def initialize(cover_state, publisher, publish_date, id = nil)
+    super(publish_date, id)
+    @cover_state = cover_state
+    @publisher = publisher
+  end
+
+  def can_be_archived
+    super || (@cover_state == 'bad')
+  end
+end
```
[the remark this answers is not in the document]
Comment (unorjikingsley) [reply classes/book/book.rb:15]: @muneebulrehman thank you for your feedback. Can you elaborate more on this? I just went through the requirement and I couldn't find anything relating to your request??

File: classes/games/game.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    super(publish_date, id)
+    @multiplayer = multiplayer
+    @last_played_at = last_played_at
+    @author = author
+  end
+
+  def can_be_archived?
+    super && Date.today.year - Date.parse(@last_played_at).year > 2
+  end
+end
```
[the remark this answers is not in the document]
Comment (unorjikingsley) [reply classes/games/game.rb:17]: @muneebulrehman Thank you for your feedback. However, all your request are not part of the requirement and I would love to create them on my free time. thank you

File: schema/schema.sql (hunk 1 above)
Comment (unorjikingsley) [reply schema/schema.sql:3]: @muneebulrehman Thank you for your feedback. However, all your request are not part of the requirement and I would love to create them on my free time. thank you

File: README.md [hunk 6]
```diff
@@ -1 +1,166 @@
...
+
+- User-Friendly Interface: The console-based interface provides a straightforward and intuitive user experience. It allows you to navigate through menus, input data, and interact with your catalog using simple commands and prompts
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LIVE DEMO -->
+
+## Walkthrough video <a name="walk-through"></a>
+
+ [video](https://drive.google.com/file/d/1TYLRgO1Jb216s9gm3s-mEgsSvbxbpiwH/view?usp=drive_link)
```
Comment (BregornOriginal) [inline README.md:58]: - [ ] Kindly record the video again, It has to be shorter, less than 5 minutes. It's not acceptable and it's not professional to see in a capstone presentation a person walking shirtless and one of the students it's not present for a couple of minutes in the video. Remember that this video could be seen by your future recruiters, and it's really important to look at professionals 🎩

File: classes/item.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
+require 'date'
+
+class Item
```
Comment (BregornOriginal) [inline classes/item.rb:3]: - [ ] As was requested by the previous reviewer and it's a requirement for this Capstone. Kindly add a test for the methods missing on this class Item

***Nowadays, the market has a lot of developers that are good at writing code, but fewer developers that write tests. Remember that besides this being a requirement for the Capstone, it's a good practice and it's really good to show your projects with a good tests suite*** 💯 💯 

image

image

**Review by BregornOriginal — CHANGES_REQUESTED** [review]
Hi @Team 

Good job so far!

There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

## Suggested changes
_Check the comments under the review._

giphy (5)

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: README.md (hunk 6 above)
Comment (unorjikingsley) [reply README.md:58]: thank you for your feedback. we would surely work on that

**Review by brainconnect93 — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi @unorjikingsley & @Paul-tes & @musangamfure, 👋 👋 

Your project is complete! Great job for following the project requirement🥇and implementing the changes requested by the previous reviewers. There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 




## Good Points 👍

- Descriptive Pull Request title & summary. ✔️
- Ruby Capstone Project completed successfully ✔️
- `Schema` file written with all table well detailed ✔️ 
- Implement all the requirement of the project ✔️ 
- Video Demonstration properly recorded ✔️ 
- Good use of Git flow ✔️

## Optional changes

- [ ] Kindly change your pull request title to something more descriptive and goes inline with what you have done in your project.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @brainconnect93 in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 2 more limited reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**